azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+
+    def _get_final_list_for_archive(self, priority_file_queue):
+        # Given a priority queue of files to collect, add one by one while the archive size is under the size limit.
+        # If a single file is over the file size limit, truncate it before adding it to the archive.
+        self._log_to_results_file("\n### Preparing list of files to add to archive ###")
+        total_uncompressed_size = 0
+        final_files_to_collect = []
+
+        while priority_file_queue:
+            file_path = heappop(priority_file_queue)[1]  # (priority, file_path)
+            file_size = min(os.path.getsize(file_path), FILE_SIZE_LIMIT)
+
+            if total_uncompressed_size + file_size > ARCHIVE_SIZE_LIMIT:
+                self._log_to_results_file("Archive too big, done with adding files.")
+                break
+
+            if os.path.getsize(file_path) <= FILE_SIZE_LIMIT:
+                final_files_to_collect.append(file_path)
+                self._log_to_results_file("Adding file {0}, size {1}b".format(file_path, file_size))
+            else:
+                truncated_file_path = self._truncate_large_file(file_path)
+                if truncated_file_path:
+                    self._log_to_results_file("Adding truncated file {0}, size {1}b".format(truncated_file_path, file_size))
+                    final_files_to_collect.append(truncated_file_path)
+
+            total_uncompressed_size += file_size
+
+        return final_files_to_collect
+
+    def _create_list_of_files_to_collect(self):
+        # The final list of files to be collected by zip is created in three steps:
+        # 1) Parse given manifest file, expanding wildcards and keeping a list of files that exist on disk
+        # 2) Assign those files a priority depending on whether they are in the must collect file list.
+        # 3) In priority order, add files to the final list to be collected, until the size of the archive is under
+        #    the size limit.
+        parsed_file_paths = self._parse_manifest_file()
+        prioritized_file_paths = self._get_priority_files_list(parsed_file_paths)
+        files_to_collect = self._get_final_list_for_archive(prioritized_file_paths)
+        return files_to_collect
+
+    def collect_logs(self):
+        """
+        Public method that collects necessary log files in a tarball that is updated each time this method is invoked.
+        The tarball is then compressed into a zip.
+        :return: Returns True if the log collection succeeded
+        """
+        try:
+            # Clear previous run's output and create base directories if they don't exist already
+            rm_files(OUTPUT_RESULTS_FILE_PATH)
+            mkdir(TRUNCATED_FILES_DIR)
+            mkdir(LOG_COLLECTOR_DIR)
+
+            files_to_collect = self._create_list_of_files_to_collect()
+            self._log_to_results_file("\n### Creating archive ###")
+
+            self._remove_deleted_files_from_archive(files_to_collect)
```

Comment:
what is the value added of the archive? i'm not sure it's buying us anything, and probably it is adding a little overhead plus extra code

Reply:
I just ran some basic perf tests and you're right - the main load is the compressing part, not the collecting files part, and since we compress every single time the collection runs, there is no benefit in keeping an intermediate archive.

Reminder, we cannot avoid compressing every collection call since `zipfile` doesn't support removing files from a compressed archive. This means we lose the ability to modify the compressed archive and make it incrementally updated, hence the need to create a compressed archive from scratch every single time.

I am changing the approach and removing the intermediate archive. Thanks for pointing it out.